Allow and implement recursive static variables.

eefriedman opened the pull request:
***Edit: Fixed now.*** I'm pretty sure the way I'm using LLVMReplaceAllUsesWith here is
unsafe... but before I figure out how to fix that, I'd like a
reality-check: is this actually useful?

rust-highfive commented:
r? @pnkfelix

(rust_highfive has picked a reviewer for you, use r? to override)

eddyb commented:
This would be really neat for `&'static` cycles, e.g. a safe doubly linked static list.

eefriedman commented:
Updated; should be safe now.  Also added an extra commit to remove the restriction on types which can't be instantiated, to allow implementing a static doubly-linked list.

alexcrichton commented:
cc @rust-lang/lang, not sure if this is a large enough change to warrant an RFC, it'd probably be nice to have a "static specification" but that may be a bit of a pipe dream.

nrc commented:
I don't think it needs an RFC and I think it is a good idea. I do think it probably deserves some eyes on it, so I created an announcement in our internals discourse: https://internals.rust-lang.org/t/heads-up-allowing-recursive-static-variables/2309

P1start commented:
Are nonsensical cycles like `static FOO: T = FOO;` and `static FOO: T = BAR; static BAR: T = FOO;` still disallowed by this PR?

Also, perhaps the tests that are deleted in this PR should instead just be moved to `run-pass`.

eefriedman commented:
@P1start The initializer for a static item can't read the contents of any static item, so `static FOO: T = FOO;` is still rejected.

Not sure the tests are particularly useful, but I guess I could move them.

eefriedman commented:
Tests re-added.

pnkfelix commented:
Surely this should at least be feature gated?

pnkfelix commented:
(that is, I can understand side-stepping the RFC process, though I'm not 100% sure I agree with doing so, even for a feature like this.  But with or without an RFC, it still seems prudent to feature-gate new functionality in the language, just so that we can kick the tires on it in nightly without worry about whether it could inadvertantly leak out into the other distribution channels.)

eefriedman commented:
`static_recursion` feature gate added.  I strongly doubt there will be any issues in practice, but better to be on the safe side, I guess.

eddyb commented:
AFAICT you had to revert the removal of `is_instantiable`. I guess it will go away once the feature gate is lifted.

bors commented:
:umbrella: The latest upstream changes (presumably #26747) made this pull request unmergeable. Please resolve the merge conflicts.

pnkfelix commented:
@eefriedman sorry I have not reviewed this yet.  I hope I will get to it this week but I am not sure if I will.

eefriedman commented:
@pnkfelix Ping.

pnkfelix commented:
@eefriedman sorry for the delay

pnkfelix commented on src/librustc_trans/trans/context.rs:
```diff
@@ -131,6 +128,11 @@ pub struct LocalCrateContext<'tcx> {
     /// Cache of closure wrappers for bare fn's.
     closure_bare_wrapper_cache: RefCell<FnvHashMap<ValueRef, ValueRef>>,
 
+    /// List of globals for static variables which need to be RAUW'ed when
```
nit: as far as I can tell the acronym "RAUW" does not occur elsewhere in our code.

So, please expand this occurrence and then include the acronym afterwards in parentheses, effectively defining the acronym.  Like so: "static variables which need to be passed through the LLVM Replace All Uses With (RAUW) functionality"

pnkfelix commented:
okay looks good as far as my knowledge of LLVM goes.

r=me after next rebase (and hopefully the noted nit is addressed).

eefriedman commented:
@pnkfelix Rebased.

pnkfelix commented:
@bors r+

bors commented:
:pushpin: Commit 0eea0f6 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 0eea0f6e907ff8c8a25a10004bede7b4621b50aa -->

bors commented:
:hourglass: Testing commit 0eea0f6 with merge e333e6a...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/5839), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/5854), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/5796), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/5825), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5801), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/5876), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/5867), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5832), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/812), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/816), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/811), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/814), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/106), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/802)
